File: verifiers/python/crucible_verify_python/schema.py
"""Field-for-field Python mirror of ``apps/verifier/pkg/testreport``.

Every JSON tag here must match the Go canonical exactly. The roundtrip
test (``tests/test_schema_roundtrip.py``) compares against a sample emitted
by the Go canonical encoder; if it drifts, both sides must move together.

We use ``dataclasses`` (not pydantic) to keep the dependency footprint
small inside the verifier sandbox and to keep ``mypy --strict`` happy.
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from datetime import UTC, datetime
from enum import StrEnum
from typing import Any, Literal

# --- Enumerations --------------------------------------------------------


class Language(StrEnum):
    PYTHON = "python"
    TYPESCRIPT = "typescript"
    RUST = "rust"
    GO = "go"
    JAVA = "java"
    SWIFT = "swift"
    POLYGLOT = "polyglot"


class Tier(StrEnum):
    MUTATION = "tier_0_mutation"
    PBT = "tier_1_pbt"
    CONTRACT = "tier_2_contract"
    PROOF = "tier_3_proof"
    HONEST_CI = "tier_4_honest_ci"


class Verdict(StrEnum):
    PASSED = "passed"
    FAILED = "failed"
    TIMED_OUT = "timed_out"
    TOOL_UNAVAILABLE = "tool_unavailable"
    SKIPPED = "skipped"


# Helper aliases for findings — keep flexible so the Go side can evolve.
Category = Literal[
    "mutation_survived",
    "property_failed",
    "fuzz_crash",
    "contract_violation",
    "proof_obligation_unmet",
    "honest_ci_mismatch",
    "tool_error",
]
Severity = Literal["info", "warn", "error"]


# --- Per-tier stats ------------------------------------------------------


@dataclass(slots=True)
class SurvivedMutant:
    file: str
    line: int
    mutator: str
    original: str = ""
    replacement: str = ""


@dataclass(slots=True)
class MutationStats:
    killed: int = 0
    survived: int = 0
    not_covered: int = 0
    timeout: int = 0
    total: int = 0
    score: float = 0.0
    threshold: float = 0.85
    diff_scoped: bool = True
    mutated_files: list[str] = field(default_factory=list)
    survived_summary: list[SurvivedMutant] = field(default_factory=list)


@dataclass(slots=True)
class Counterexample:
    property: str
    shrunk: str
    seed: str = ""
    stack_hint: str = ""


@dataclass(slots=True)
class PBTStats:
    iterations: int = 0
    iterations_min: int = 10_000
    properties: list[str] = field(default_factory=list)
    counterexamples: list[Counterexample] = field(default_factory=list)
    fuzz_corpus_size: int = 0
    fuzz_new_seeds: int = 0
    fuzz_crashes: int = 0


@dataclass(slots=True)
class ContractViolation:
    endpoint: str
    method: str
    check: str
    detail: str
    reproducer: str = ""


@dataclass(slots=True)
class ContractStats:
    spec_path: str = ""
    spec_hash: str = ""
    stateful_workflows: int = 0
    checks: list[str] = field(default_factory=list)
    violations: list[ContractViolation] = field(default_factory=list)
    dst_iterations: int = 0
    dst_replay_id: str = ""
    dst_failing_schedule: str = ""


@dataclass(slots=True)
class ProofStats:
    prover: str = ""
    proof_artifact: str = ""
    obligations: int = 0
    discharged: int = 0
    timed_out: bool = False
    wall_clock_seconds: float = 0.0
    cached_partial: bool = False
    fallback_tier: str = ""
    codeowner_review_required: bool = False
    unsoundness_hints: list[str] = field(default_factory=list)


@dataclass(slots=True)
class HonestCIStats:
    builder_id: str = ""
    nix_flake_hash: str = ""
    nix_lock_hash: str = ""
    executor_rebuild_hash: str = ""
    verifier_rebuild_hash: str = ""
    bit_identical: bool = False
    slsa_level: int = 0
    in_toto_statement_hash: str = ""
    fulcio_cert_hash: str = ""
    rekor_uuid: str = ""
    witness_attestation: str = ""
    tekton_chains_ref: str = ""
    diffoscope_report: str = ""
    scrubber_audit_ok: bool = False
    scrubber_audit_entries: int = 0


@dataclass(slots=True)
class Finding:
    category: str
    severity: str
    detail: str
    file: str = ""
    line: int = 0
    suggested_fix: str = ""


# --- The report ----------------------------------------------------------


# Fields that are zero-valued (Go's omitempty equivalent) on the Go side
# and so MUST be elided from our JSON output to keep canonical hashes
# stable. These mirror the `,omitempty` JSON tags in testreport.go.
_OMITEMPTY_KEYS: frozenset[str] = frozenset(
    {
        "tool_digest",
        "reporter_version",
        "reporter_oidc_subject",
        "error",
        "findings",
        "mutation",
        "pbt",
        "contract",
        "proof",
        "honest_ci",
    }
)
_MUTATION_OMITEMPTY: frozenset[str] = frozenset(
    {"not_covered", "timeout", "mutated_files", "survived_summary"}
)
_PBT_OMITEMPTY: frozenset[str] = frozenset(
    {"properties", "counterexamples", "fuzz_corpus_size", "fuzz_new_seeds", "fuzz_crashes"}
)
_CONTRACT_OMITEMPTY: frozenset[str] = frozenset(
    {
        "spec_path",
        "spec_hash",
        "stateful_workflows",
        "checks",
        "violations",
        "dst_iterations",
        "dst_replay_id",
        "dst_failing_schedule",
    }
)
_PROOF_OMITEMPTY: frozenset[str] = frozenset(
    {
        "proof_artifact",
        "obligations",
        "discharged",
        "wall_clock_seconds",
        "cached_partial",
        "fallback_tier",
        "codeowner_review_required",
        "unsoundness_hints",
    }
)
_HONEST_OMITEMPTY: frozenset[str] = frozenset(
    {
        "nix_flake_hash",
        "nix_lock_hash",
        "in_toto_statement_hash",
        "fulcio_cert_hash",
        "rekor_uuid",
        "witness_attestation",
        "tekton_chains_ref",
        "diffoscope_report",
        "scrubber_audit_entries",
    }
)
_FINDING_OMITEMPTY: frozenset[str] = frozenset({"file", "line", "suggested_fix"})
_SURVIVED_OMITEMPTY: frozenset[str] = frozenset({"original", "replacement"})
_CONTRACT_VIOLATION_OMITEMPTY: frozenset[str] = frozenset({"reproducer"})
_COUNTEREXAMPLE_OMITEMPTY: frozenset[str] = frozenset({"seed", "stack_hint"})


def _is_zero(value: Any) -> bool:
    """Return True if value should be elided under Go's omitempty rule."""
    if value is None:
        return True
    if isinstance(value, str | list | dict) and len(value) == 0:
        return True
    if isinstance(value, bool):
        return value is False
    if isinstance(value, int | float):
        return value == 0
    return False


def _strip_omitempty(d: dict[str, Any], omit: frozenset[str]) -> dict[str, Any]:
    return {k: v for k, v in d.items() if not (k in omit and _is_zero(v))}


@dataclass(slots=True)
class TestReport:
    """Mirrors :go:`testreport.TestReport`.

    Required fields are at the top; optional/per-tier substructs follow.
    The schema version is pinned — bumps require a 90-day deprecation.
    """

    # Tell pytest not to collect this class as a test container.
    __test__ = False

    task_id: str
    tier: Tier
    language: Language = Language.PYTHON
    framework: str = ""
    verdict: Verdict = Verdict.SKIPPED
    passed: bool = False
    schema_version: str = "1"
    diff_hash: str = ""
    started_at: datetime = field(default_factory=lambda: datetime.now(tz=UTC))
    finished_at: datetime = field(default_factory=lambda: datetime.now(tz=UTC))
    duration_seconds: float = 0.0
    wall_clock_budget_seconds: float = 0.0
    mutation: MutationStats | None = None
    pbt: PBTStats | None = None
    contract: ContractStats | None = None
    proof: ProofStats | None = None
    honest_ci: HonestCIStats | None = None
    findings: list[Finding] = field(default_factory=list)
    tool_digest: str = ""
    reporter_id: str = "crucible-verify-python"
    reporter_version: str = ""
    reporter_oidc_subject: str = ""
    error: str = ""

    # --- serialisation ----------------------------------------------------

    def to_json_dict(self) -> dict[str, Any]:
        """Render the report as a plain dict whose shape matches Go's encoder.

        Field order mirrors the Go struct field order so that human-eyeball
        diffs against a Go-emitted sample line up cleanly. Hashing
        consumers must canonicalise themselves (sorted keys + no
        whitespace) — :func:`canonical_json` does so.
        """
        # Pull values explicitly to preserve Go field order.
        out: dict[str, Any] = {
            "schema_version": self.schema_version,
            "task_id": self.task_id,
            "diff_hash": self.diff_hash,
            "tier": str(self.tier),
            "language": str(self.language),
            "framework": self.framework,
            "verdict": str(self.verdict),
            "passed": self.passed,
            "started_at": _format_time(self.started_at),
            "finished_at": _format_time(self.finished_at),
            "duration_seconds": self.duration_seconds,
            "wall_clock_budget_seconds": self.wall_clock_budget_seconds,
        }
        if self.mutation is not None:
            out["mutation"] = _stats_to_dict_mutation(self.mutation)
        if self.pbt is not None:
            out["pbt"] = _stats_to_dict_pbt(self.pbt)
        if self.contract is not None:
            out["contract"] = _stats_to_dict_contract(self.contract)
        if self.proof is not None:
            out["proof"] = _stats_to_dict_proof(self.proof)
        if self.honest_ci is not None:
            out["honest_ci"] = _stats_to_dict_honest(self.honest_ci)
        if self.findings:
            out["findings"] = [_finding_to_dict(f) for f in self.findings]
        if self.tool_digest:
            out["tool_digest"] = self.tool_digest
        out["reporter_id"] = self.reporter_id
        if self.reporter_version:
            out["reporter_version"] = self.reporter_version
        if self.reporter_oidc_subject:
            out["reporter_oidc_subject"] = self.reporter_oidc_subject
        if self.error:
            out["error"] = self.error
        return out

    def to_json(self) -> str:
        """Render the report as a UTF-8 JSON string (compact)."""
        return json.dumps(self.to_json_dict(), separators=(",", ":"), ensure_ascii=False)

    def canonical_json(self) -> str:
        """Sorted-keys, compact JSON suitable for hashing/attestation."""
        return json.dumps(
            self.to_json_dict(), separators=(",", ":"), sort_keys=True, ensure_ascii=False
        )

    # --- invariants -------------------------------------------------------

    def validate(self) -> None:
        """Enforce the same invariants as :go:`TestReport.Validate`."""
        if self.schema_version != "1":
            raise ValueError(
                f"testreport: schema_version {self.schema_version!r} != '1'"
            )
        if not self.task_id:
            raise ValueError("testreport: task_id required")
        if not str(self.tier):
            raise ValueError("testreport: tier required")
        if not str(self.language):
            raise ValueError("testreport: language required")
        if self.duration_seconds < 0:
            raise ValueError("testreport: negative duration")
        if self.tier == Tier.MUTATION and self.mutation is not None:
            if not self.mutation.diff_scoped:
                raise ValueError(
                    "testreport: mutation report not diff-scoped (Crucible mandate)"
                )
            if (
                self.mutation.total > 0
                and self.mutation.killed + self.mutation.survived > self.mutation.total
            ):
                raise ValueError("testreport: mutation killed+survived > total")
        if (
            self.tier == Tier.PBT
            and self.pbt is not None
            and self.pbt.iterations_min > 0
            and self.pbt.iterations < self.pbt.iterations_min
        ):
            raise ValueError(
                f"testreport: PBT iterations {self.pbt.iterations} "
                f"< required {self.pbt.iterations_min}"
            )


# --- internal helpers ----------------------------------------------------


def _format_time(t: datetime) -> str:
    """Match Go's ``time.Time`` JSON encoder (RFC 3339 with ns precision).

    Go renders trailing-zero-trimmed nanos. Python only carries microsecond
    precision via ``datetime``, so the encoder emits at most 6 fractional
    digits. The dispatcher is tolerant of either format.
    """
    if t.tzinfo is None:
        t = t.replace(tzinfo=UTC)
    # isoformat() yields e.g. "2026-05-15T12:34:56.789012+00:00".
    # Go's default uses "Z" suffix for UTC; normalise for round-trip parity.
    s = t.astimezone(UTC).isoformat()
    if s.endswith("+00:00"):
        s = s[: -len("+00:00")] + "Z"
    return s


def _stats_to_dict_mutation(m: MutationStats) -> dict[str, Any]:
    d = asdict(m)
    d["survived_summary"] = [
        _strip_omitempty(asdict(s), _SURVIVED_OMITEMPTY) for s in m.survived_summary
    ]
    return _strip_omitempty(d, _MUTATION_OMITEMPTY)


def _stats_to_dict_pbt(p: PBTStats) -> dict[str, Any]:
    d = asdict(p)
    d["counterexamples"] = [
        _strip_omitempty(asdict(c), _COUNTEREXAMPLE_OMITEMPTY) for c in p.counterexamples
    ]
    return _strip_omitempty(d, _PBT_OMITEMPTY)


def _stats_to_dict_contract(c: ContractStats) -> dict[str, Any]:
    d = asdict(c)
    d["violations"] = [
        _strip_omitempty(asdict(v), _CONTRACT_VIOLATION_OMITEMPTY) for v in c.violations
    ]
    return _strip_omitempty(d, _CONTRACT_OMITEMPTY)


def _stats_to_dict_proof(p: ProofStats) -> dict[str, Any]:
    return _strip_omitempty(asdict(p), _PROOF_OMITEMPTY)


def _stats_to_dict_honest(h: HonestCIStats) -> dict[str, Any]:
    return _strip_omitempty(asdict(h), _HONEST_OMITEMPTY)


def _finding_to_dict(f: Finding) -> dict[str, Any]:
    return _strip_omitempty(asdict(f), _FINDING_OMITEMPTY)


# Re-exports for callers that want to introspect the omitempty contracts
# without touching the underscored helpers.
__all__ = [
    "ContractStats",
    "ContractViolation",
    "Counterexample",
    "Finding",
    "HonestCIStats",
    "Language",
    "MutationStats",
    "PBTStats",
    "ProofStats",
    "SurvivedMutant",
    "TestReport",
    "Tier",
    "Verdict",
]

File: verifiers/rust/src/schema.rs
//! Schema mirror for `apps/verifier/pkg/testreport/testreport.go`.
//!
//! Field naming, ordering, and `omitempty` semantics are reproduced
//! exactly: all JSON keys are `snake_case` and every optional field that
//! Go marks with `,omitempty` is modelled as `Option<T>` plus
//! `#[serde(skip_serializing_if = "Option::is_none")]` (or, for
//! collections, `Vec::is_empty` / `default+skip_serializing_if`). The
//! dispatcher unmarshals the resulting JSON straight into the Go struct,
//! so any drift here will surface as a hard schema error in
//! `processpool.parseRunnerOutput`.

#![allow(missing_docs)]

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Schema contract version. Bumped only on breaking change (90-day
/// deprecation per `twin-spec/schemas/README.md`).
pub const SCHEMA_VERSION: &str = "1";

/// In-toto `predicateType` URI for every TestReport.
pub const PREDICATE_TYPE: &str = "https://crucible.dev/TestReport/v1";

/// Reporter identifier baked into emitted reports.
pub const REPORTER_ID: &str = "crucible-verify-rust";
/// Reporter version baked into emitted reports.
pub const REPORTER_VERSION: &str = env!("CARGO_PKG_VERSION");

/// Per-language runner identity (mirrors `testreport.Language`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Language {
    Python,
    Typescript,
    Rust,
    Go,
    Java,
    Swift,
    Polyglot,
}

impl Default for Language {
    fn default() -> Self {
        Self::Rust
    }
}

/// Tier identifier (mirrors `testreport.Tier`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Tier {
    #[serde(rename = "tier_0_mutation")]
    Mutation,
    #[serde(rename = "tier_1_pbt")]
    Pbt,
    #[serde(rename = "tier_2_contract")]
    Contract,
    #[serde(rename = "tier_3_proof")]
    Proof,
    #[serde(rename = "tier_4_honest_ci")]
    HonestCi,
}

impl Tier {
    /// Parse the CLI tier flag (e.g. `"tier_0_mutation"`).
    pub fn parse(s: &str) -> anyhow::Result<Self> {
        match s {
            "tier_0_mutation" => Ok(Self::Mutation),
            "tier_1_pbt" => Ok(Self::Pbt),
            "tier_2_contract" => Ok(Self::Contract),
            "tier_3_proof" => Ok(Self::Proof),
            "tier_4_honest_ci" => Ok(Self::HonestCi),
            other => anyhow::bail!("unknown tier {other:?}"),
        }
    }

    /// Wire string identical to the Go constant.
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Mutation => "tier_0_mutation",
            Self::Pbt => "tier_1_pbt",
            Self::Contract => "tier_2_contract",
            Self::Proof => "tier_3_proof",
            Self::HonestCi => "tier_4_honest_ci",
        }
    }
}

/// Runner-local pass/fail signal (mirrors `testreport.Verdict`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Verdict {
    Passed,
    Failed,
    TimedOut,
    ToolUnavailable,
    Skipped,
}

impl Default for Verdict {
    fn default() -> Self {
        Self::Skipped
    }
}

/// Single runner+tier+language report — the wire shape the Go dispatcher
/// unmarshals. Order is best-effort identical to the Go struct (serde
/// preserves declaration order for struct serialisation).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TestReport {
    pub schema_version: String,
    pub task_id: String,
    pub diff_hash: String,
    pub tier: Tier,
    pub language: Language,
    pub framework: String,
    pub verdict: Verdict,
    pub passed: bool,

    pub started_at: DateTime<Utc>,
    pub finished_at: DateTime<Utc>,
    pub duration_seconds: f64,
    pub wall_clock_budget_seconds: f64,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub mutation: Option<MutationStats>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pbt: Option<PbtStats>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub contract: Option<ContractStats>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub proof: Option<ProofStats>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub honest_ci: Option<HonestCiStats>,

    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub findings: Vec<Finding>,

    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub tool_digest: String,

    pub reporter_id: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub reporter_version: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub reporter_oidc_subject: String,

    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub error: String,
}

impl TestReport {
    /// Build a freshly-stamped report for `(tier, language)` with the
    /// reporter identity already baked in.
    pub fn new(tier: Tier, language: Language, task_id: &str, diff_hash: &str) -> Self {
        let now = Utc::now();
        Self {
            schema_version: SCHEMA_VERSION.to_string(),
            task_id: task_id.to_string(),
            diff_hash: diff_hash.to_string(),
            tier,
            language,
            framework: String::new(),
            verdict: Verdict::Skipped,
            passed: false,
            started_at: now,
            finished_at: now,
            duration_seconds: 0.0,
            wall_clock_budget_seconds: 0.0,
            mutation: None,
            pbt: None,
            contract: None,
            proof: None,
            honest_ci: None,
            findings: Vec::new(),
            tool_digest: String::new(),
            reporter_id: REPORTER_ID.to_string(),
            reporter_version: REPORTER_VERSION.to_string(),
            reporter_oidc_subject: String::new(),
            error: String::new(),
        }
    }

    /// Mark the report as `tool_unavailable` with a human-readable detail.
    pub fn tool_unavailable(mut self, detail: impl Into<String>) -> Self {
        self.verdict = Verdict::ToolUnavailable;
        self.passed = false;
        self.error = detail.into();
        self
    }

    /// Stamp `finished_at` and `duration_seconds` from the supplied start.
    pub fn stamp_finished(&mut self) {
        let now = Utc::now();
        self.finished_at = now;
        let dur = now
            .signed_duration_since(self.started_at)
            .num_milliseconds()
            .max(0) as f64
            / 1000.0;
        self.duration_seconds = dur;
    }
}

/// Tier-0 mutation statistics (cargo-mutants).
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct MutationStats {
    pub killed: u32,
    pub survived: u32,
    #[serde(default, skip_serializing_if = "is_zero_u32")]
    pub not_covered: u32,
    #[serde(default, skip_serializing_if = "is_zero_u32")]
    pub timeout: u32,
    pub total: u32,
    pub score: f64,
    pub threshold: f64,
    pub diff_scoped: bool,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub mutated_files: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub survived_summary: Vec<SurvivedMutant>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct SurvivedMutant {
    pub file: String,
    pub line: u32,
    pub mutator: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub original: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub replacement: String,
}

/// Tier-1 property test statistics (proptest, cargo-fuzz).
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PbtStats {
    pub iterations: u64,
    pub iterations_min: u64,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub properties: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub counterexamples: Vec<Counterexample>,
    #[serde(default, skip_serializing_if = "is_zero_u64")]
    pub fuzz_corpus_size: u64,
    #[serde(default, skip_serializing_if = "is_zero_u64")]
    pub fuzz_new_seeds: u64,
    #[serde(default, skip_serializing_if = "is_zero_u64")]
    pub fuzz_crashes: u64,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Counterexample {
    pub property: String,
    pub shrunk: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub seed: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub stack_hint: String,
}

/// Tier-2 contract statistics (schemathesis).
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ContractStats {
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub spec_path: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub spec_hash: String,
    #[serde(default, skip_serializing_if = "is_zero_u32")]
    pub stateful_workflows: u32,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub checks: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub violations: Vec<ContractViolation>,
    #[serde(default, skip_serializing_if = "is_zero_u64")]
    pub dst_iterations: u64,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub dst_replay_id: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub dst_failing_schedule: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ContractViolation {
    pub endpoint: String,
    pub method: String,
    pub check: String,
    pub detail: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub reproducer: String,
}

/// Tier-3 proof statistics (Kani).
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ProofStats {
    pub prover: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub proof_artifact: String,
    #[serde(default, skip_serializing_if = "is_zero_u32")]
    pub obligations: u32,
    #[serde(default, skip_serializing_if = "is_zero_u32")]
    pub discharged: u32,
    pub timed_out: bool,
    #[serde(default, skip_serializing_if = "is_zero_f64")]
    pub wall_clock_seconds: f64,
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub cached_partial: bool,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub fallback_tier: String,
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub codeowner_review_required: bool,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub unsoundness_hints: Vec<String>,
}

/// Tier-4 honest-CI statistics.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct HonestCiStats {
    pub builder_id: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub nix_flake_hash: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub nix_lock_hash: String,
    pub executor_rebuild_hash: String,
    pub verifier_rebuild_hash: String,
    pub bit_identical: bool,
    pub slsa_level: u32,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub in_toto_statement_hash: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub fulcio_cert_hash: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub rekor_uuid: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub witness_attestation: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub tekton_chains_ref: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub diffoscope_report: String,
    pub scrubber_audit_ok: bool,
    #[serde(default, skip_serializing_if = "is_zero_u32")]
    pub scrubber_audit_entries: u32,
}

/// Structured finding — mirrors `testreport.Finding`. The rubric LLM-judge
/// folds these into `VerifierRejection.RejectionReasons`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Finding {
    pub category: String,
    pub severity: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub file: String,
    #[serde(default, skip_serializing_if = "is_zero_u32")]
    pub line: u32,
    pub detail: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub suggested_fix: String,
}

/// Subset of the `verification.VerificationRequest` Go struct relevant to
/// the per-language Rust runner. We only deserialise the fields the
/// runner consumes; unknown fields are tolerated by `serde(default)` and
/// `serde_json::Value` for the diff body.
#[derive(Debug, Clone, Default, Deserialize)]
pub struct VerificationRequest {
    #[serde(default)]
    pub task_id: String,
    #[serde(default)]
    pub tenant_id: String,
    #[serde(default)]
    pub repo: String,
    #[serde(default)]
    pub base_sha: String,
    #[serde(default)]
    pub diff: Diff,
    #[serde(default)]
    pub test_files: Vec<FileChange>,
    #[serde(default)]
    pub spec_changes: Vec<SpecChange>,
    #[serde(default)]
    pub languages: Vec<String>,
    #[serde(default)]
    pub executor_sandbox_id: String,
    #[serde(default)]
    pub budget: BudgetEnvelope,
}

#[derive(Debug, Clone, Default, Deserialize)]
pub struct Diff {
    #[serde(default)]
    pub files: Vec<FileChange>,
}

#[derive(Debug, Clone, Default, Deserialize)]
pub struct FileChange {
    #[serde(default)]
    pub path: String,
    #[serde(default)]
    pub status: String,
    #[serde(default)]
    pub old_path: String,
    #[serde(default)]
    pub unified_diff: String,
    #[serde(default)]
    pub before_blob_sha: String,
    #[serde(default)]
    pub after_blob_sha: String,
}

#[derive(Debug, Clone, Default, Deserialize)]
pub struct SpecChange {
    #[serde(default)]
    pub path: String,
    #[serde(default)]
    pub kind: String,
    #[serde(default)]
    pub previous_hash: String,
    #[serde(default)]
    pub current_hash: String,
    #[serde(default)]
    pub delta: String,
}

#[derive(Debug, Clone, Default, Deserialize)]
pub struct BudgetEnvelope {
    #[serde(default)]
    pub verifier_cap_usd: f64,
    #[serde(default)]
    pub verifier_spent_usd: f64,
    #[serde(default)]
    pub wall_clock_cap_seconds: u64,
    #[serde(default)]
    pub wall_clock_spent_seconds: u64,
}

// ----- serde helpers -------------------------------------------------------

#[allow(clippy::trivially_copy_pass_by_ref)]
fn is_zero_u32(v: &u32) -> bool {
    *v == 0
}

#[allow(clippy::trivially_copy_pass_by_ref)]
fn is_zero_u64(v: &u64) -> bool {
    *v == 0
}

#[allow(clippy::trivially_copy_pass_by_ref)]
fn is_zero_f64(v: &f64) -> bool {
    *v == 0.0
}
